net: coap: Fix response matching algorithm

rlubos opened the pull request:
* Add test case for matching pending replies with received responses.
Cover corner cases that are failing with the current implementation.

* Fix response matching algorithm 

Fixes #84100

rlubos committed:
tests: net: coap: Add test case for matching pending replies

Add test case for matching pending replies with received responses.
Cover corner cases that are failing with the current implementation.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: coap: Fix response matching algorithm

The algorithm for matching request with response was incorrect, which
could lead to false matches (for example if request had a token, and
piggybacked reply had no token but matching message ID only, that would
still be counted as a match).

This commit fixes it. The request/reply matching is implemented based on
RFC now, with separate conditions for piggybacked/separate responses.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

boaks commented on subsys/net/lib/coap/coap.c:
```diff
@@ -1817,23 +1819,44 @@ struct coap_reply *coap_response_received(
 	tkl = coap_header_get_token(response, token);
 
+	if (type == COAP_TYPE_ACK || type == COAP_TYPE_RESET) {
+		piggybacked = true;
```
A RST is always a empty message and no piggybacked response.

See [RFC 7252, page 23, table 1](https://www.rfc-editor.org/rfc/rfc7252#page-23)

```
+----------+-----+-----+-----+-----+
|          | CON | NON | ACK | RST |
+----------+-----+-----+-----+-----+
| Request  | X   | X   | -   | -   |
| Response | X   | X   | X   | -   |
| Empty    | *   | -   | X   | X   |
+----------+-----+-----+-----+-----+
  Table 1: Usage of Message Types
```

boaks commented on subsys/net/lib/coap/coap.c:
```diff
@@ -1817,23 +1819,44 @@ struct coap_reply *coap_response_received(
+				if (memcmp(r->token, token, r->tkl) != 0) {
+					continue;
+				}
+			}
```
```
if (piggybacked) {
   if ((r->id != id) {
      continue;
   }   
}

if ((r->tkl != tkl) {
   continue;
}   

if (r->tkl > 0) {
   if (memcmp(r->token, token, r->tkl) != 0) {
      continue;
   }
}
```

will use less code-duplicates.

rlubos replied:
Thanks for the suggestion, I took the wrong assumption that for a separate response matching token means non-empty, which apparently is not the case. Simplified the code as suggested, and updated relevant test cases.

boaks commented on subsys/net/lib/coap/coap.c:
```diff
@@ -1817,23 +1819,44 @@ struct coap_reply *coap_response_received(
 		return NULL;
 	}
```
```
	if (!is_empty_message(response) && coap_packet_is_request(response)) {
		/* Request can't be response */
		return NULL;
	}
```
The **!** before `is_empty_message` is wrong. Both requests and empty messages can't be responses, see table below.

rlubos replied:
Good point, that check indeed seems wrong. Changed to `if (is_empty_message(response) || coap_packet_is_request(response))` and added test cases to cover empty reply and request case.

boaks commented on subsys/net/lib/coap/coap.c:
```diff
@@ -1817,23 +1819,44 @@ struct coap_reply *coap_response_received(
 	for (i = 0, r = replies; i < len; i++, r++) {
 		int age;
 
 		if ((r->id == 0U) && (r->tkl == 0U)) {
```
Is the message ID 0 used to indicate something special in the zephyr implementation?
if not, what is the purpose of this check?

FMPOV, it's wrong. I don't see this is related to RFC 7252. 
Considering a separate response with empty token, assumptions about the used MIDs will fail. The other implementation may use 0 without considering any special processing, as I don't see one in RFC 7252

rlubos replied:
Can't tell for the original author, but it seems this was just the way to skip unused entries in the array. As the current approach is not compliant with the RFC, I've changed this to check if the reply callback is set instead.

boaks commented on tests/net/lib/coap/src/main.c:
```diff
@@ -1855,4 +1855,87 @@ ZTEST(coap, test_age_is_newer)
+		{ .id = 102, .type = COAP_TYPE_ACK, .match = NULL,
+		  .token = { 1, 2, 3, 4 }, .tkl = 4 },
+		/* Separate reply, no token */
+		{ .id = 100, .type = COAP_TYPE_CON, .match = NULL },
```
For separate replies, there is no relation in the MIDs!
Using different MIDs for this test, e.g. 1000, would show that.

boaks commented on tests/net/lib/coap/src/main.c:
```diff
@@ -1855,4 +1855,87 @@ ZTEST(coap, test_age_is_newer)
+		/* Separate reply, no token */
+		{ .id = 100, .type = COAP_TYPE_CON, .match = NULL },
+		/* Separate reply, no token 2 */
+		{ .id = 101, .type = COAP_TYPE_CON, .match = NULL },
```
separate response, MID must be ignored, empty token matches first reply.

boaks commented on subsys/net/lib/coap/coap.c:
```diff
@@ -1817,23 +1819,44 @@ struct coap_reply *coap_response_received(
+			/* In a separate response, just the tokens of the
+			 * response and original request MUST match.
+			 */
+			if ((r->tkl == 0) || (r->tkl != tkl) ||
```
If the code-duplicates are intended, then the `(r->tkl == 0)` is wrong and must be removed. That will then also change the results of the tests, as I commended there.

rlubos replied:
Reworked as suggested in earlier comment.

rlubos commented:
@boaks Thanks for the detailed review, please revisit.

Aluleichtbau commented:
We continued the discussion in https://github.com/zephyrproject-rtos/zephyr/issues/84100. I added my test there, but @rlubos is right: It is better to post them here. In https://github.com/zephyrproject-rtos/zephyr/issues/84100#issuecomment-2601970742 I realized there is still an issue with it. So here is the fixed version of the tests. You may just want to append them to `tests/net/lib/coap/src/main.c` check against that.

```c
ZTEST(coap, test_response_received)
{
	int r;
	struct coap_packet response = {0};
	static const char payload[] = "payload";
	uint8_t *data = data_buf[0];
	struct sockaddr from;
	struct coap_reply replies[] = {
		{.id = 0, .token = "", .tkl = 0, .reply = NULL},
		{.id = 0, .token = "", .tkl = 0, .reply = response_handled_cb},
		{.id = 0, .token = "no mid", .tkl = 6, .reply = response_handled_cb},
		{.id = 123, .token = "", .tkl = 0, .reply = response_handled_cb},
		{.id = 124, .token = "token", .tkl = 5, .reply = response_handled_cb},
		{.id = 125, .token = "token", .tkl = 5, .reply = response_handled_cb},
		{.id = 125, .token = "other", .tkl = 5, .reply = response_handled_cb},
		{.id = 126, .token = "token", .tkl = 3, .reply = response_handled_cb},
	};

	memset(data_buf[0], 0, ARRAY_SIZE(data_buf[0]));

	/* empty ACK for a reply without token*/
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_ACK, 0, NULL,
			     COAP_CODE_EMPTY, 123);
	zassert_equal(r, 0, "Could not initialize packet");
	zassert_is_null(coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
			"A reply is picked");

	/* empty ACK for a reply with a token*/
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_ACK, 0, NULL,
			     COAP_CODE_EMPTY, 124);
	zassert_equal(r, 0, "Could not initialize packet");
	zassert_is_null(coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
			"A reply is picked");

	/* empty ACK for a reply with mid = 0*/
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_ACK, 0, NULL,
			     COAP_CODE_EMPTY, 0);
	zassert_equal(r, 0, "Could not initialize packet");
	zassert_is_null(coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
			"A reply is picked");

	/* empty ACK could not be matched*/
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_ACK, 0, NULL,
			     COAP_CODE_EMPTY, 1);
	zassert_equal(r, 0, "Could not initialize packet");
	zassert_is_null(coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
			"A reply is picked");

	/* piggybacked ACK without token matched */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_ACK, 0, NULL,
			     COAP_RESPONSE_CODE_CONTENT, 123);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_equal_ptr(
		coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
		&replies[3], "Wrong reply is picked");

	/* piggybacked ACK without token not matched */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_ACK, 0, NULL,
			     COAP_RESPONSE_CODE_CONTENT, 124);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_is_null(coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
			"A reply is picked");

	/* piggybacked ACK with token matched */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_ACK,
			     strlen("token"), "token", COAP_RESPONSE_CODE_CONTENT, 124);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_equal_ptr(
		coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
		&replies[4], "Wrong reply is picked");

	/* piggybacked ACK with token not matched - unknown id and token */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_ACK,
			     strlen("idk"), "idk", COAP_RESPONSE_CODE_CONTENT, 1);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_is_null(coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
			"A reply is picked");

	/* piggybacked ACK with token not matched - token exists, mid exists but not combination */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_ACK,
			     strlen("other"), "other", COAP_RESPONSE_CODE_CONTENT, 124);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_is_null(coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
			"A reply is picked");

	/* piggybacked ACK with token not matched - wrong token len */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_ACK,
			     strlen("token"), "token", COAP_RESPONSE_CODE_CONTENT, 126);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_is_null(coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
			"A reply is picked");

	/* CON without token matched - id equal */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_CON, 0, NULL,
			     COAP_RESPONSE_CODE_CONTENT, 123);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_equal_ptr(
		coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
		&replies[3], "Wrong reply is picked");

	/* CON without token not matched - id not found */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_CON, 0, NULL,
			     COAP_RESPONSE_CODE_CONTENT, 1);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_is_null(coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
			"A reply is picked");

	/* CON without token not matched - id lacks token */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_CON, 0, NULL,
			     COAP_RESPONSE_CODE_CONTENT, 124);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_is_null(coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
			"A reply is picked");

	/* CON with token matched */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_CON,
			     strlen("token"), "token", COAP_RESPONSE_CODE_CONTENT, 124);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_equal_ptr(
		coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
		&replies[4], "Wrong reply is picked");

	/* CON with token not matched - unknown id and token */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_CON,
			     strlen("idk"), "idk", COAP_RESPONSE_CODE_CONTENT, 1);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_is_null(coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
			"A reply is picked");

	/* CON with token matched - mid differs, but token matches */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_CON,
			     strlen("no mid"), "no mid", COAP_RESPONSE_CODE_CONTENT, 0);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_equal_ptr(
		coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
		&replies[2], "Wrong reply is picked");

	/* NON_CON without token matched - id equal */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_NON_CON, 0,
			     NULL, COAP_RESPONSE_CODE_CONTENT, 123);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_equal_ptr(
		coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
		&replies[3], "Wrong reply is picked");

	/* NON_CON without token not matched - id not found */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_NON_CON, 0,
			     NULL, COAP_RESPONSE_CODE_CONTENT, 1);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_is_null(coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
			"A reply is picked");

	/* NON_CON without token not matched - id lacks token */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_NON_CON, 0,
			     NULL, COAP_RESPONSE_CODE_CONTENT, 124);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_is_null(coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
			"A reply is picked");

	/* NON_CON with token matched */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_NON_CON,
			     strlen("token"), "token", COAP_RESPONSE_CODE_CONTENT, 124);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_equal_ptr(
		coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
		&replies[4], "Wrong reply is picked");

	/* NON_CON with token not matched - unknown id and token */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_NON_CON,
			     strlen("idk"), "idk", COAP_RESPONSE_CODE_CONTENT, 1);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_is_null(coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
			"A reply is picked");

	/* NON_CON with token matched - mid differs, but token matches */
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_NON_CON,
			     strlen("no mid"), "no mid", COAP_RESPONSE_CODE_CONTENT, 0);
	zassert_equal(r, 0, "Could not initialize packet");

	r = coap_packet_append_payload_marker(&response);
	zassert_equal(r, 0, "Failed to set the payload marker");

	r = coap_packet_append_payload(&response, payload, sizeof(payload));
	zassert_equal(r, 0, "Failed to set the payload");

	zassert_equal_ptr(
		coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
		&replies[2], "Wrong reply is picked");

	/* empty RESET for a reply without token*/
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_RESET, 0,
			     NULL, COAP_CODE_EMPTY, 123);
	zassert_equal(r, 0, "Could not initialize packet");
	zassert_equal_ptr(
		coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
		&replies[3], "Wrong reply is picked");

	/* empty RESET for a reply with a token*/
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_RESET, 0,
			     NULL, COAP_CODE_EMPTY, 124);
	zassert_equal(r, 0, "Could not initialize packet");
	zassert_equal_ptr(
		coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
		&replies[4], "Wrong reply is picked");

	/* empty RESET for a reply with mid = 0*/
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_RESET, 0,
			     NULL, COAP_CODE_EMPTY, 0);
	zassert_equal(r, 0, "Could not initialize packet");
	zassert_equal_ptr(
		coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
		&replies[1], "Wrong reply is picked");

	/* empty RESET could not be matched*/
	r = coap_packet_init(&response, data, COAP_BUF_SIZE, COAP_VERSION_1, COAP_TYPE_RESET, 0,
			     NULL, COAP_CODE_EMPTY, 1);
	zassert_equal(r, 0, "Could not initialize packet");
	zassert_is_null(coap_response_received(&response, &from, &replies[0], ARRAY_SIZE(replies)),
			"A reply is picked");
}
```

boaks commented on tests/net/lib/coap/src/main.c:
```diff
@@ -1855,4 +1855,110 @@ ZTEST(coap, test_age_is_newer)
+	};
+	struct test_coap_request test_responses[] = {
+		/* #0 Piggybacked ACK, no token */
+		{ .id = 100, .type = COAP_TYPE_ACK, .match = &matches[1],
```
Maybe too picky, but RFC 7252 named that "empty token" not "no token". In the case of an empty message (code 0.0) the it may be considered as "no token", but for non empty messages the wording is an empty token.

rlubos replied:
Changed to "empty token"

rlubos committed:
net: coap: Fix coap_packet_is_request() check for empty code

Empty code was incorrectly matched as a request, fix that.

Align coap_handle_request_len() function to behave as documented in the
API documentation - in case of invalid request code (which is also the
case for empty code) -ENOTSUP Should be returned.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
@Aluleichtbau Thanks for the test cases, I've extended the test list in this PR and I think we're pretty much covered now. Two caveats though:
* a few of your test cases wrongly assumed that for a separate response MID has to match,
* I've skipped tests for NON CON, IMO they're redundant as from reply matching perspective there's not really a difference between CON and NON CON, both are treated as separate replies

Aluleichtbau commented:
> @Aluleichtbau Thanks for the test cases, I've extended the test list in this PR and I think we're pretty much covered now. Two caveats though:
> 
>     * a few of your test cases wrongly assumed that for a separate response MID has to match,
> 
>     * I've skipped tests for NON CON, IMO they're redundant as from reply matching perspective there's not really a difference between CON and NON CON, both are treated as separate replies

You are right. There was still an issue with them, I updated them. NON_CON and CON is indeed equal. I copied them just for sake of completeness, but removing NON_CON is also fine. Thank you!

boaks commented on subsys/net/lib/coap/coap.c:
```diff
@@ -1172,7 +1172,7 @@ bool coap_packet_is_request(const struct coap_packet *cpkt)
 	uint8_t code = coap_header_get_code(cpkt);
 
-	return !(code & ~COAP_REQUEST_MASK);
+	return (code != COAP_CODE_EMPTY) && !(code & ~COAP_REQUEST_MASK);
```
There is a "ping/pong" definition in RFC 7252 (CON 0.0 => RST 0.0), where, depending on the implementation, an empty CON message may be considered to be a request. 
Not sure, if that is related, but that change may fail with a ping/pong.

rlubos replied:
I was tempted to change this initially, but TBH after digging into RFC 7252 I'm not sure if we should consider such an empty CON "probe" as a request. Specifically, chapter 4.1 defines an empty message separately from request/response:
> Separate from the message type, a message may carry a request, a response, or be Empty.

Similarily in chapter 4.2 and empty message is considered separately from regular request/response:
> A Confirmable message always carries either a request or response, unless it is used only to elicit a Reset message, in which case it is Empty.

Of course if the above assumption is not a valid one I guess we could just add an extra type check to the function instead.

boaks replied:
The one view is the wording, and with that the change is well.
The other view is the processing of such an CON 0.0.
I'm not sure, what the logic inside the zephyr functions for that ping/pong is. I'm still on NCS 2.6.2 so I don't see the current implementation (I would need to check that out to see, how such a ping pong is intended to be processed. I will go for that tomorrow.)

boaks commented:
From my side LGTM.

pdgendt commented:
CI appears to have issues?

rlubos commented:
I see they've merged some Clang-related hotfix yesterday evening, let's try a rerun.
